### DISCUSSION TITLE
Background indexing not performed when using Emacs Eglot and clangd

### DISCUSSION THREAD

**Discussion by mikaelspringer** [clangd/clangd#1748] [Q&A]
I use clangd through Emacs and Eglot. I have a `.clangd` configuration file in the project directory with the following entry;
```
Index:
   Background: Build
```
and, I also start clangd with a number of flags as can be seen in the Eglot events log;
```
[stderr] I[10:37:29.103] Homebrew clangd version 16.0.6
[stderr] I[10:37:29.104] Features: mac+xpc
[stderr] I[10:37:29.104] PID: 49201
[stderr] I[10:37:29.104] Working directory: /Users/mikael/Develop/reason
[stderr] I[10:37:29.104] argv[0]: /opt/homebrew/opt/llvm/bin/clangd
[stderr] I[10:37:29.104] argv[1]: --enable-config
[stderr] I[10:37:29.104] argv[2]: --clang-tidy
[stderr] I[10:37:29.104] argv[3]: --log=info
[stderr] I[10:37:29.104] argv[4]: --pretty
[stderr] I[10:37:29.104] argv[5]: --background-index
[stderr] I[10:37:29.104] argv[6]: -j=4
[stderr] I[10:37:29.104] argv[7]: --completion-style=detailed
[stderr] I[10:37:29.105] Starting LSP over stdin/stdout
```
When I open a source file I can see it is indexed (in the Emacs mini-buffer). If I try to find all references to an item in the source file I only see the references in the current source file. But I know there are a lot of references in other source files so I open one of them. This source file is in turn indexed. After this, if I repeat finding all references of the item I can now see references in the second source file as well. I interpret this as files being indexed when they are opened. But isn't the background index option meant to make clangd index all files present in the compile_commands.json (which I forgot to mention I have in the project root)? This should mean all references should be visible right away and not after opening each source file. Or have I misunderstood this?

**Comment by HighCommander4** [comment]
> But isn't the background index option meant to make clangd index all files present in the compile_commands.json (which I forgot to mention I have in the project root)? This should mean all references should be visible right away and not after opening each source file. Or have I misunderstood this?

Your understanding is correct.

Has background indexing for the project completed? It can take a while for large projects, and while it's in progress find-refs results will be incomplete. Clangd infroms the client about the progress of background indexing, though I'm not sure where emacs surfaces this.

If that's not the issue, something else is going wrong. A full log with `--log=verbose` might contain some clues.
